op-challenger: Asterisc Support with Refactoring

pcw109550 opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

TL; DR: Implemented op-challenger for asterisc with refactoring. Only unit test is passing. E2E tests will be implemented at asterisc repo. 

**Description**

op-challenger provides a wrapper with FPVM binary, and it only supported cannon to play dispute game, providing traces. This PR adds a wrapper with yet another FPVM, [asterisc](https://github.com/ethereum-optimism/asterisc).

Initial implementation(https://github.com/ethereum-optimism/optimism/pull/10030) of op-challenger for asterisc had some duplicated codes, and even cyclic dependency from the nature that asterisc is maintained at separate repository. 

This PR removes asterisc dependencies (removing code duplicates) by asking asterisc binary to already include the witness and stateHash in the state json. Related asterisc fixes: https://github.com/ethereum-optimism/asterisc/pull/37, https://github.com/ethereum-optimism/asterisc/pull/36. These FPVM fixes may be also added to cannon, to decouple FPVM and the challenger. It is FPVM's job to provide correct witness and stateHash.

I tried to avoid duplicate codes by exporting cannon tracer methods as much as possible. These shared logics can be refactored out. op-challenger for cannon is currently running on testnet, and I wanted to avoid code changes for cannon in this PR.

**Refactoring Suggestion**

##### FPVM abstraction

FPVM maintains its [state](https://github.com/ethereum-optimism/specs/blob/a9a5d15e358eefd3a1303f4187754a64728a5709/specs/experimental/fault-proof/cannon-fault-proof-vm.md#state), and witness and stateHash. In the FPVM state, `step`, `PC`, `Exited` will be always be included, whatever FPVM we choose. Therefore, we may use below struct to parse state json from any FPVM.

```go
type VMState struct {
	PC        uint64   `json:"pc"`
	Exited    bool     `json:"exited"`
	Step      uint64   `json:"step"`
	Witness   []byte   `json:"witness"`
	StateHash [32]byte `json:"stateHash"`
}
```

Upper fields are the only required data to perform op-challenger logic. I was able to implement asterisc tracer based on upper fields. 

op-challenger can perform basic validation of this data. It can check the witness length since this may be differ by the VM type. Also, we may validate `exitCode` which is included at `StateHash`.

##### Metricer

Current `Metricer` interface requires `RecordCannonExecutionTime()` and `RecordAsteriscExecutionTime()` to be implemented. I first thought to refactor these methods to `RecordFPVMExecutionTime()`, but it may break compatibility for prometheus. 

##### Flags

Most of the flags for asterisc binary is duplicated from cannon: (See https://github.com/ethereum-optimism/asterisc/pull/8), and we may think all the duplicated flag codes([op-challenger/flags/flags.go](https://github.com/ethereum-optimism/optimism/pull/10094/commits/76677015a037f280c12af624ca18ed7b1094e1e5#diff-bebb5fc1cf1aaf7bcabd7f512a9354b8980ed7dc3cbc8b98196c8aca07bcb668)) are unnecessary. However, I thought these flags are not included in the specs, so they are subject to change. 

**Tests**

Unit tests for asterisc tracer is added. Basic flag sanity checking unit test is added.

There are no e2e tests added! These will be added to [asterisc](https://github.com/ethereum-optimism/asterisc). Will refer to [output_cannon_test.go](https://github.com/ethereum-optimism/optimism/blob/develop/op-e2e/faultproofs/output_cannon_test.go).

To add more details about why e2e is implemented at asterisc, asterisc is expected to be maintained at separate repository. Asterisc pulls the monorepo(here!)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Added support for Asterisc trace type configurations and flags.
	- Introduced a new game type, `Asterisc`, in the system.
	- Implemented functionality to parse and validate Asterisc VM states.
	- Added creation and access capabilities for traces related to the Asterisc fault type.
	- Introduced new metrics to record execution times for Asterisc operations.

- **Refactor**
	- Renamed several functions and data structures for consistency and clarity in the Cannon trace type modules.

- **Bug Fixes**
	- Updated test functions to properly handle new configurations and fault types.

- **Documentation**
	- Updated method and variable declarations to include new configurations and functionalities for Asterisc.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pcw109550 committed:
op-challenger: Refactor cannon trace provider

Move gzip method to ioutil

pcw109550 committed:
op-challenger: Expose cannon tracer methods

pcw109550 committed:
op-challenger: asterisc implementation

Basic unit tests mirrored from cannon.

pcw109550 committed:
op-challenger: asterisc flags and configs

pcw109550 committed:
op-challenger: asterisc metric helper

pcw109550 committed:
op-challenger: asterisc registration and add types

pcw109550 committed:
op-dispute-mon: enable asterisc type

pcw109550 commented:
@coderabbitai review

pcw109550 committed:
Fix typo

refcell commented on op-challenger/game/fault/trace/cannon/executor.go:
```diff
@@ -157,9 +157,9 @@ func runCmd(ctx context.Context, l log.Logger, binary string, args ...string) er
+// FindStartingSnapshot finds the closest snapshot before the specified traceIndex in snapDir.
 // If no suitable snapshot can be found it returns absolutePreState.
-func findStartingSnapshot(logger log.Logger, snapDir string, absolutePreState string, traceIndex uint64) (string, error) {
+func FindStartingSnapshot(logger log.Logger, snapDir string, absolutePreState string, traceIndex uint64) (string, error) {
```
I see we're using this in a few packages outside `cannon`, would this be a good candidate to refactor out as well?

pcw109550 replied:
Created a `utils` package. Refactored out at https://github.com/ethereum-optimism/optimism/pull/10094/commits/d536958da6ca05b2f7e2ba2e4f3b5456d78d8d7f, https://github.com/ethereum-optimism/optimism/pull/10094/commits/ffc3f1fc8b79b16d24f4f2e06dfdc4b0dc7933d3.

refcell commented on op-challenger/game/fault/trace/cannon/executor.go:
```diff
@@ -141,11 +141,11 @@ func (e *Executor) generateProof(ctx context.Context, dir string, begin uint64,
 }
 
-func runCmd(ctx context.Context, l log.Logger, binary string, args ...string) error {
+func RunCmd(ctx context.Context, l log.Logger, binary string, args ...string) error {
```
Likewise with `RunCmd`, is this something we can refactor into maybe a trace utils package?

refcell replied:
I'm thinking we could update the cannon executor file to more closely mirror the simplicity in the asterisc executor by refactoring these standalone methods out.

pcw109550 replied:
Yep. Removed all the cannon op-challenger module dependency from asterisc.

refcell commented on op-challenger/cmd/main_test.go:
```diff
@@ -29,6 +29,12 @@ var (
 	datadir                 = "./test_data"
 	cannonL2                = "http://example.com:9545"
 	rollupRpc               = "http://example.com:8555"
+	asteriscNetwork         = "op-mainnet"
```
Let's de-dup this and just use `testNetwork` and then `otherTestNetwork`

refcell commented on op-challenger/config/config.go:
```diff
@@ -35,16 +35,30 @@ var (
 	ErrCannonNetworkUnknown          = errors.New("unknown cannon network")
 	ErrMissingRollupRpc              = errors.New("missing rollup rpc url")
 )
+var (
+	ErrMissingAsteriscL2               = errors.New("missing asterisc L2")
+	ErrMissingAsteriscBin              = errors.New("missing asterisc bin")
+	ErrMissingAsteriscServer           = errors.New("missing asterisc server")
+	ErrMissingAsteriscAbsolutePreState = errors.New("missing asterisc absolute pre-state")
+	ErrMissingAsteriscSnapshotFreq     = errors.New("missing asterisc snapshot freq")
+	ErrMissingAsteriscInfoFreq         = errors.New("missing asterisc info freq")
+	ErrMissingAsteriscRollupConfig     = errors.New("missing asterisc network or rollup config path")
+	ErrMissingAsteriscL2Genesis        = errors.New("missing asterisc network or l2 genesis path")
+	ErrAsteriscNetworkAndRollupConfig  = errors.New("only specify one of network or rollup config path")
+	ErrAsteriscNetworkAndL2Genesis     = errors.New("only specify one of network or l2 genesis path")
+	ErrAsteriscNetworkUnknown          = errors.New("unknown asterisc network")
+)
```
Let's combine these variable blocks, but we can divide them with an empty line for readability.

```suggestion

	ErrMissingAsteriscL2               = errors.New("missing asterisc L2")
	ErrMissingAsteriscBin              = errors.New("missing asterisc bin")
	ErrMissingAsteriscServer           = errors.New("missing asterisc server")
	ErrMissingAsteriscAbsolutePreState = errors.New("missing asterisc absolute pre-state")
	ErrMissingAsteriscSnapshotFreq     = errors.New("missing asterisc snapshot freq")
	ErrMissingAsteriscInfoFreq         = errors.New("missing asterisc info freq")
	ErrMissingAsteriscRollupConfig     = errors.New("missing asterisc network or rollup config path")
	ErrMissingAsteriscL2Genesis        = errors.New("missing asterisc network or l2 genesis path")
	ErrAsteriscNetworkAndRollupConfig  = errors.New("only specify one of network or rollup config path")
	ErrAsteriscNetworkAndL2Genesis     = errors.New("only specify one of network or l2 genesis path")
	ErrAsteriscNetworkUnknown          = errors.New("unknown asterisc network")
)
```

pcw109550 replied:
Fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/ceb1b24643cb3cc9c50f7fc1c9b90f15096a729c.

refcell commented on op-challenger/config/config_test.go:
```diff
@@ -24,6 +24,14 @@ var (
 	validRollupRpc             = "http://localhost:8555"
 )
 
+var (
```
nit: merge these blocks. Can delineate with an empty line between for legibility.

```suggestion

```

refcell commented on op-challenger/config/config_test.go:
```diff
@@ -24,6 +24,14 @@ var (
 	validRollupRpc             = "http://localhost:8555"
 )
 
+var (
+	validAsteriscBin             = "./bin/asterisc"
```
nit: merge these blocks. Can delineate with an empty line between for legibility.

```suggestion

	validAsteriscBin             = "./bin/asterisc"
```

pcw109550 replied:
Fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/ceb1b24643cb3cc9c50f7fc1c9b90f15096a729c.

refcell commented on op-challenger/config/config_test.go:
```diff
@@ -209,7 +224,7 @@ func TestRollupRpcRequired(t *testing.T) {
 }
 
-func TestRequireConfigForMultipleTraceTypes(t *testing.T) {
+func TestRequireConfigForMultipleTraceTypesForCannon(t *testing.T) {
 	cfg := validConfig(TraceTypeCannon)
 	cfg.TraceTypes = []TraceType{TraceTypeCannon, TraceTypeAlphabet}
```
Should we also be testing **both** cannon and alphabet trace types? i.e. if a challenger wants to run against both cannon and asterisc?

It's likely in production that multiple honest actors would be run for isolated trace types to avoid resource contention but it'd be good to test and assert that this is the expected behavior for the `op-challenger`'s config and it behaves according to expectation.

pcw109550 replied:
> Should we also be testing **both** cannon and alphabet trace types?

Are you saying that we should test the case when we use both cannon and asterisc, not both cannon and alphabet(This already exists at `TestRequireConfigForMultipleTraceTypesForCannon`)?

Like 
```go
func TestRequireConfigForMultipleTraceTypesForCannonAndAsterisc(t *testing.T) {
	cfg1 := validConfig(TraceTypeAsterisc)
        cfg2 := validConfig(TraceTypeCannon)
        cfg := mergeConfig(cfg1, cfg2)
        // and go on flag validation
```

pcw109550 replied:
Added test at https://github.com/ethereum-optimism/optimism/pull/10094/commits/838d7d6e7b23d9ec7bad95fbeaf3c390a3de8df4.

refcell commented on op-challenger/game/fault/register.go:
```diff
@@ -58,7 +59,7 @@ func RegisterGameTypes(
 	var closer CloseFunc
 	var l2Client *ethclient.Client
-	if cfg.TraceTypeEnabled(config.TraceTypeCannon) || cfg.TraceTypeEnabled(config.TraceTypePermissioned) {
+	if cfg.TraceTypeEnabled(config.TraceTypeCannon) || cfg.TraceTypeEnabled(config.TraceTypePermissioned) || cfg.TraceTypeEnabled(config.TraceTypeAsterisc) {
```
We'll need to put the asterisc dialing in a separate block because this dials the `cfg.CannonL2`, but the asterisc flags allow for a different l2 url to be dialed.

refcell replied:
Alternatively, within this scope, we can set variably declare the url to dial based on the trace type.

e.g.
```golang
l2RpcUrl := if cfg.TraceTypeEnabled(config.TraceTypeAsterisc) { cfg.AsteriscL2 } else { cfg.CannonL2 }
```

refcell commented on op-challenger/game/fault/trace/asterisc/asterisc_state.go:
```diff
@@ -0,0 +1,69 @@
+package asterisc
```
nit: Can we remove the `asterisc_` prefix from this file and the `asterisc_state_test.go` file since they're already within the `asterisc` directory/package?

pcw109550 replied:
Done at https://github.com/ethereum-optimism/optimism/pull/10094/commits/dd3ad27496887b74b39865181b8ea5091d90f429. Also fixed `cannon_state.go` and `cannon_state_test.go` at https://github.com/ethereum-optimism/optimism/pull/10094/commits/dd3ad27496887b74b39865181b8ea5091d90f429.

refcell commented on op-challenger/game/fault/trace/asterisc/asterisc_state_test.go:
```diff
@@ -0,0 +1,83 @@
+//go:embed test_data/state.json
+var testState []byte
+
+func TestLoadState(t *testing.T) {
```
Nice test coverage here 👍

refcell commented on op-challenger/game/fault/trace/asterisc/asterisc_state.go:
```diff
@@ -0,0 +1,69 @@
+var asteriscWitnessLen = 362
+
+// The state struct will be read from json.
+// other fields included in json are specific to FPVM implementation, and not required for tracer
```
nit: trace provider
```suggestion
// other fields included in json are specific to FPVM implementation, and not required for trace provider
```

pcw109550 replied:
Fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/f11fa4cd784c6769b0a0da161e3cbf2becfa01c9.

refcell commented on op-challenger/game/fault/trace/asterisc/executor.go:
```diff
@@ -0,0 +1,134 @@
+	return e.generateProof(ctx, dir, i, i)
+}
+
+// generateProofOrUntilPreimageRead executes asterisc to generate a proof at the specified trace index,
```
nit: fix doc comment function name
```suggestion
// generateProof executes asterisc to generate a proof at the specified trace index,
```

refcell commented on op-challenger/game/fault/trace/asterisc/executor.go:
```diff
@@ -0,0 +1,134 @@
+	return e.generateProof(ctx, dir, i, i)
+}
+
+// generateProofOrUntilPreimageRead executes asterisc to generate a proof at the specified trace index,
+// or until a non-local preimage read is encountered if untilPreimageRead is true.
+// The proof is stored at the specified directory.
```
nit: fix doc comment. I see this is incorrect in the upstream cannon executor `generateProof` method docs. Will fix that in a separate pr.
```suggestion
// generateProof executes asterisc from the specified starting trace index until the end trace index.
// The proof is stored at the specified directory.
```

pcw109550 replied:
Fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/f11fa4cd784c6769b0a0da161e3cbf2becfa01c9. Also went ahead and fixed cannon's comment at https://github.com/ethereum-optimism/optimism/pull/10094/commits/864a7970a824d644e91d6822317e296ac87aa40d.

refcell commented on op-challenger/game/fault/trace/asterisc/prestate_test.go:
```diff
@@ -0,0 +1,59 @@
+	})
+
+	t.Run("CacheAbsolutePreState", func(t *testing.T) {
+		setupPreState(t, dataDir, "state.json")
```
nit: use the specified `prestate` variable.
```suggestion
		setupPreState(t, dataDir, prestate)
```

pcw109550 replied:
Fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/ceb1b24643cb3cc9c50f7fc1c9b90f15096a729c#diff-bbcfab6ff345d2048d0e9f697f4c233f22cd0af92995ba6c0088c49971aa64b9. Cannon test had identical issue so fixed at https://github.com/ethereum-optimism/optimism/pull/10094/commits/0d44932a7da1209b9053e8625fbc8537d2df1dd5.

refcell commented on op-challenger/game/fault/trace/asterisc/provider.go:
```diff
@@ -0,0 +1,256 @@
+
+func (p *AsteriscTraceProvider) Get(ctx context.Context, pos types.Position) (common.Hash, error) {
+	traceIndex := pos.TraceIndex(p.gameDepth)
+	if !traceIndex.IsUint64() {
+		return common.Hash{}, errors.New("trace index out of bounds")
+	}
+	proof, err := p.loadProof(ctx, traceIndex.Uint64())
```
We'll need to expand this to a `uint128` to support the expanded trace index bounds. Will follow up sync in our meeting today.

pcw109550 replied:
I understood that we do not have to expand this, because traceIndex at this point is already safe, not overflowing over Uint64. There is no cannon trace provider fix as well.

refcell commented on op-challenger/game/fault/trace/asterisc/provider.go:
```diff
@@ -0,0 +1,256 @@
+
+func (p *AsteriscTraceProvider) GetStepData(ctx context.Context, pos types.Position) ([]byte, []byte, *types.PreimageOracleData, error) {
+	traceIndex := pos.TraceIndex(p.gameDepth)
+	if !traceIndex.IsUint64() {
+		return nil, nil, nil, errors.New("trace index out of bounds")
+	}
+	proof, err := p.loadProof(ctx, traceIndex.Uint64())
```
ibid - here too.

pcw109550 replied:
ditto.

refcell commented on op-challenger/game/fault/trace/asterisc/provider.go:
```diff
@@ -0,0 +1,256 @@
+	logger         log.Logger
+	dir            string
+	prestate       string
+	generator      cannon.ProofGenerator
```
Yea so it'd be nice here to pull these out so the `asterisc` package has no dependency on the `cannon` package.

refcell commented on op-challenger/game/fault/trace/asterisc/provider.go:
```diff
@@ -0,0 +1,256 @@
+	logger         log.Logger
+	dir            string
+	prestate       string
+	generator      cannon.ProofGenerator
+	gameDepth      types.Depth
+	preimageLoader *cannon.PreimageLoader
```
Yea so it'd be nice here to pull these (`ProofGenerator` and `PreimageLoader`) out so the `asterisc` package has no dependency on the `cannon` package.

pcw109550 replied:
Modularized out.

refcell reviewed (commented):
Looking very good. Mostly nits, a few minor fixes.

pcw109550 committed:
Fix l2RPC when registering asterisc

pcw109550 committed:
Deduplicate preimageOpts for testing

pcw109550 committed:
Fix comments

pcw109550 committed:
Style issue fix

pcw109550 committed:
Deduplicate network flag for testing

pcw109550 committed:
Remove redundant prefix at filename

refcell reviewed (commented):
Changes look great! Last conversations to resolve then we should be good to merge

pcw109550 committed:
Add config test for using both cannon and asterisc

pcw109550 committed:
Refactor asterisc/cannon provider using utils

pcw109550 committed:
Remove redundant prefix at filename

pcw109550 committed:
Fix comments

pcw109550 committed:
Style issue fix

pcw109550 committed:
Refactor asterisc/cannon provider using utils

pcw109550 commented on op-challenger/game/fault/trace/utils/executor.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+// FindStartingSnapshot finds the closest snapshot before the specified traceIndex in snapDir.
+// If no suitable snapshot can be found it returns absolutePreState.
+func FindStartingSnapshot(logger log.Logger, snapDir string, absolutePreState string, traceIndex uint64) (string, error) {
+	// Find the closest snapshot to start from
+	entries, err := os.ReadDir(snapDir)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return absolutePreState, nil
+		}
+		return "", fmt.Errorf("list snapshots in %v: %w", snapDir, err)
+	}
+	bestSnap := uint64(0)
+	for _, entry := range entries {
+		if entry.IsDir() {
+			logger.Warn("Unexpected directory in snapshots dir", "parent", snapDir, "child", entry.Name())
+			continue
+		}
+		name := entry.Name()
+		if !snapshotNameRegexp.MatchString(name) {
+			logger.Warn("Unexpected file in snapshots dir", "parent", snapDir, "child", entry.Name())
+			continue
+		}
+		index, err := strconv.ParseUint(name[0:len(name)-len(".json.gz")], 10, 64)
+		if err != nil {
+			logger.Error("Unable to parse trace index of snapshot file", "parent", snapDir, "child", entry.Name())
+			continue
+		}
+		if index > bestSnap && index < traceIndex {
+			bestSnap = index
+		}
+	}
+	if bestSnap == 0 {
+		return absolutePreState, nil
+	}
+	startFrom := fmt.Sprintf("%v/%v.json.gz", snapDir, bestSnap)
+
+	return startFrom, nil
+}
```
Avoid racing by checking by single system call. We still need to check err of `os.ReadDir`.

pcw109550 commented:
@refcell I removed asterisc's trace provider dependency of cannon by creating new `utils` module. Also fixed nits(commits, style issues) and added suggested tests.

refcell reviewed (approved):
This looks great! Would appreciate another proofs-reviewer

ajsutton commented on op-challenger/game/fault/register.go:
```diff
@@ -58,10 +60,14 @@ func RegisterGameTypes(
-	if cfg.TraceTypeEnabled(config.TraceTypeCannon) || cfg.TraceTypeEnabled(config.TraceTypePermissioned) {
-		l2, err := ethclient.DialContext(ctx, cfg.CannonL2)
+	if cfg.TraceTypeEnabled(config.TraceTypeCannon) || cfg.TraceTypeEnabled(config.TraceTypePermissioned) || cfg.TraceTypeEnabled(config.TraceTypeAsterisc) {
+		l2RpcUrl := cfg.CannonL2
+		if cfg.TraceTypeEnabled(config.TraceTypeAsterisc) {
+			l2RpcUrl = cfg.AsteriscL2
+		}
```
I don't think this is right since we might have both cannon and asterisc enabled at the same time.  I don't really see a reason to support different L2s for cannon and asterisc so maybe we should just change `--cannon-l2` to be just `--l2` (with a `--cannon-l2` alias for backwards compatibility) and use it for both.

ajsutton commented on op-challenger/game/fault/trace/asterisc/provider.go:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+func (p *AsteriscTraceProvider) SetMaxDepth(gameDepth types.Depth) {
+	p.gameDepth = gameDepth
+}
```
This seems to be unused and can be removed.  Feel free to remove it from the cannon provider while you're at it. :)

ajsutton reviewed (commented):
Looks good generally. Couple of nits and I think we need to fix up the confusion over having two different L2 URL options.

pcw109550 committed:
Remove unused methods

ajsutton commented on op-challenger/flags/flags.go:
```diff
@@ -71,6 +71,11 @@ var (
 		EnvVars: prefixEnvVars("MAX_CONCURRENCY"),
 		Value:   uint(runtime.NumCPU()),
 	}
+	L2RpcFlag = &cli.StringFlag{
+		Name:    "l2-rpc",
```
My suggestion would be to add:
```suggestion
	L2RpcFlag = &cli.StringFlag{
		Name:    "l2-rpc",
        Aliases: []string{"cannon-l2"},
```

and remove the `CannonL2Flag`. Means we don't have to manage the two separate flags.

ajsutton reviewed (approved):
LGTM. Nice work.

ajsutton commented:
@pcw109550 if you're happy with this I think we're good to merge.

pcw109550 commented:
@ajsutton I think its time to merge! Can you put this PR into the merge queue? Thanks!

ajsutton commented:
Ah, this broke the fpac-devnet tests because my suggestion to use an alias for `--cannon-l2` was a dumb idea as it turns out.  While it does work to have `--cannon-l2`, it no longer supports the env var `OP_CHALLENGER_CANNON_L2` which we use in that devnet but also in a production setup.

So sorry but can we go back to it being a separate option again?

ajsutton commented:
@pcw109550 See above, sorry.

pcw109550 committed:
op-challenger: add l2-rpc flag and deprecate cannon-l2 flag

pcw109550 commented:
@ajsutton No problem. I have gone back to having separate options again. Added tests. Also manually checked that the binary works like below.
If we provide both flags: `--l2-rpc` and `--cannon-l2`. 
```
./bin/op-challenger  --l2-rpc 2 --cannon-l2 3 ...
CRIT [04-19|15:48:34.828] Application failed                       err="failed to setup: flag cannon-l2 and l2-rpc must not be both set"
```
If both are not provided,
```
CRIT [04-19|15:52:17.435] Application failed                       err="failed to setup: flag l2-rpc is required"
```
If at least one between `--l2-rpc` and `--cannon-l2` is provided, it works fine.
